frontend/model/notifications/utils.js

```diff
@@ -121,3 +120,14 @@ export function isOlder (notification: Notification): boolean {
 export function maxAge (notification: Notification): number {
   return notification.read ? MAX_AGE_READ : MAX_AGE_UNREAD
 }
+
+export function makeNotificationHash (notification: Object): string {
+  const excludingKeys = ['read', 'hash', 'body']
```

Comment:
Hmm, by excluding the `body` key, is it possible that two different notifications might get hashed the same way? We want to make sure the text of the notification is hashed. If it's stored in the `body` key, it probably shouldn't be excluded.

Reply:
> by excluding the body key, is it possible that two different notifications might get hashed the same way?

Simply, the `body` is the outcome of translation. 😁 So we should not include it.

##
For the `uniqueConsistentID`, it's a good idea. And I don't think we don't need to add `type` in the `uniqueConsistentID` because the `notification` object has its type`.
```
const notification = {
  ...template,
  hash: '',
  avatarUserID: template.avatarUserID || sbp('state/vuex/getters').ourIdentityContractId,
  // Sets 'groupID' if this notification only pertains to a certain group.
  ...(template.scope === 'group' ? { groupID: data.groupID } : {}),
  // Store integer timestamps rather than ISO strings here to make age comparisons easier.
  timestamp: data.createdDate ? new Date(data.createdDate).getTime() : Date.now(),
  type
}
notification.hash = makeNotificationHash(notification)
```
And for the notifications created inside the contract sideEffect, I think there are no problems with the current function `makeNotificationHash` because it uses `meta.createdDate` as `createdDate` of notification itself.
But for the other notifications created outside of contract sideEffect, their `createdDate` could differ from one device to another even though they are same notifications. So, I think we should consider only this one.

Reply:
However, I totally agree with your idea that we should use `uniqueConsistentID`.
I thought you wanted to only add that field and still using `makeNotificationHash` function. But if we add `uniqueConsistentID`, we don't need `makeNotificationHash` function anymore. 😁👍

Reply:
@taoeffect, while I was working on this, I noticed that we don't need `uniqueConsistentID` and we can still have identifiable data to make hash using `makeNotificationHash`. The only reason why we need `uniqueConsistentID` was that we set `createdDate` inside the `notificationData` in the place outside of contract sideEffect.
```
sbp('gi.notifications/emit', 'NEAR_DISTRIBUTION_END', {
  createdDate: new Date().toISOString(),
  groupID: rootState.currentGroupId,
  period: rootGetters.groupSettings.distributionDate
})
```
But we don't need to set the `createdDate` here because we have a way to use the default value `Date.now()`.
```
const notification = {
  ...template,
  hash: makeNotificationHash({ ...data, type }),
  avatarUserID: template.avatarUserID || sbp('state/vuex/getters').ourIdentityContractId,
  // Sets 'groupID' if this notification only pertains to a certain group.
  ...(template.scope === 'group' ? { groupID: data.groupID } : {}),
  // Store integer timestamps rather than ISO strings here to make age comparisons easier.
  timestamp: data.createdDate ? new Date(data.createdDate).getTime() : Date.now(),
  type
}
```
This means that we can easily using `makeNotificationHash` by removing `createdDate` in the place where we pass `notificationData`.

And to be honest, I really prefer to use `makeNotificationHash` because to make ID (`uniqueConsistentID`) manually is something like tricky temporary solution when we don't have the better solution. We also create hash for GIMessages using the identifiable value.
```
hash = createCID(value)
```

Reply:
Removing `createdDate` from the hashed data only addresses **Problem 2** in the comment I wrote above.

**Problem 1** is still an issue. So I think we need to introduce `uniqueConsistentID`, unless you have some other idea.

Reply:
I don't think there is a problem (Problem 1), because notificationData contains `createdDate`, which is `meta.createdDate`, when the we create notifications in the contract sideEffect.

Reply:
Notification is created by it's type and it's data.
```
'gi.notifications/emit' (type: string, data: NotificationData) {
  const template: NotificationTemplate = templates[type](data)

  if (template.scope === 'group' && !data.groupID) {
    throw new TypeError('Incomplete notification data: `data.groupID` is required.')
  }

  // Creates the notification object in a single step.
  const notification = {
    ...template,
    hash: makeNotificationHash({ ...data, type }),
    avatarUserID: template.avatarUserID || sbp('state/vuex/getters').ourIdentityContractId,
    // Sets 'groupID' if this notification only pertains to a certain group.
    ...(template.scope === 'group' ? { groupID: data.groupID } : {}),
    // Store integer timestamps rather than ISO strings here to make age comparisons easier.
    timestamp: data.createdDate ? new Date(data.createdDate).getTime() : Date.now(),
    type
  }
  sbp('gi.actions/identity/kv/addNotificationStatus', notification)
  sbp('state/vuex/commit', keys.ADD_NOTIFICATION, notification)
}
```
For the notifications which are created in contracts sideEffect, their type and data are same in any devices.

Reply:
> I don't think there is a problem (Problem 1), because notificationData contains createdDate, which is meta.createdDate, when the we create notifications in the contract sideEffect.

There is a problem. It is possible for notifications to contain the same data but be different notifications.

For example, `PAYMENT_THANKYOU_SENT` has this:

```js
  PAYMENT_THANKYOU_SENT (data: { creatorID: string, fromMemberID: string, toMemberID: string }) {
    return {
      avatarUserID: data.fromMemberID,
      body: L('{strong_}{name}{_strong} sent you a {strong_}thank you note{_strong} for your contribution.', {
        name: `${CHATROOM_MEMBER_MENTION_SPECIAL_CHAR}${data.fromMemberID}`,
        ...LTags('strong')
      }),
      creatorID: data.fromMemberID,
      icon: '',
      level: 'info',
      linkTo: `/payments?modal=ThankYouNoteModal&from=${data.fromMemberID}&to=${data.toMemberID}`,
      scope: 'group'
    }
  },
```

This notification will be sent multiple times, and it will have the same exact hash each time it is sent between these two members.

Reply:
We need to focus on the data here, where we call the `gi.notifications/emit` function.
```
sbp('gi.contracts/group/emitNotificationsAfterSyncing', [contractID, innerSigningContractID], [{
  notificationName: 'PAYMENT_THANKYOU_SENT',
  notificationData: {
    createdDate: meta.createdDate,
    groupID: contractID,
    fromMemberID: innerSigningContractID,
    toMemberID: data.toMemberID
  }
}])
```

Reply:
This is going to generate the same hash every time:

```js
        sbp('gi.notifications/emit', 'NEW_DISTRIBUTION_PERIOD', {
          groupID: rootState.currentGroupId,
          creatorID: rootGetters.ourIdentityContractId,
          memberType: rootGetters.ourGroupProfile.incomeDetailsType === 'pledgeAmount' ? 'pledger' : 'receiver'
        })
```

Reply:
If you really don't want to use `uniqueConsistentID` you will have to manually perform the same checks on the `data` that's passed in to `gi.notifications/emit`.

Also, in this case please add a comment above the definition of `gi.notifications/emit` explaining all of this.

My recommendation remains to prefer `uniqueConsistentID` over `makeNotificationHash`, simply because using `uniqueConsistentID` forces the developer to think about this problem, whereas with `makeNotificationHash` the developer is allowed to ignore it, but then passes on that burden to the reviewer who has to double-check their work and also not miss any mistakes the developer made in either running afoul of **Problem 1 - same hash for different notifications** or **Problem 2 - different hash for the same notifications**.

Either way, please document both problems 1 and 2 in a comment above `gi.notifications/emit`. 🙏

Reply:
Yeah, thanks.